mesonbuild/modules/gnome.py

```diff
@@ -898,21 +896,83 @@ def gdbus_codegen(self, state, args, kwargs):
+
+                output = namebase + '-docbook'
+                custom_kwargs = {'input': xml_file,
+                                 'output': output,
+                                 'command': docbook_cmd,
+                                 'build_by_default': build_by_default
+                                 }
+                targets.append(build.CustomTarget(output, state.subdir, state.subproject, custom_kwargs))
         else:
-            self._print_gdbus_warning()
-            cmd += ['--generate-c-code', '@OUTDIR@/' + namebase, '@INPUT@']
-        outputs = [namebase + '.c', namebase + '.h']
-        custom_kwargs = {'input': xml_file,
-                         'output': outputs,
-                         'command': cmd
-                         }
-        if 'build_by_default' in kwargs:
-            custom_kwargs['build_by_default'] = kwargs['build_by_default']
-        ct = build.CustomTarget(target_name, state.subdir, state.subproject, custom_kwargs)
-        return ModuleReturnValue(ct, [ct])
+            # https://git.gnome.org/browse/glib/commit/?id=ee09bb704fe9ccb24d92dd86696a0e6bb8f0dc1a
+            if mesonlib.version_compare(self._get_native_glib_version(state), '>= 2.51.3'):
+                cmd += ['--output-directory', '@OUTDIR@']
+                code_cmd = cmd + ['--generate-c-code', namebase, '@INPUT@']
+            else:
+                self._print_gdbus_warning()
+                code_cmd = cmd + ['--generate-c-code', '@OUTDIR@/' + namebase, '@INPUT@']
+
+            output = namebase + '.c'
+            custom_kwargs = {'input': xml_file,
+                             'output': output,
+                             'command': code_cmd,
+                             'build_by_default': build_by_default
+                             }
+            targets.append(build.CustomTarget(output, state.subdir, state.subproject, custom_kwargs))
+
+            output = namebase + '.h'
+            custom_kwargs = {'input': xml_file,
+                             'output': output,
+                             'command': code_cmd,
+                             'build_by_default': build_by_default
+                             }
+            targets.append(build.CustomTarget(output, state.subdir, state.subproject, custom_kwargs))
+
+            if 'docbook' in kwargs:
+                docbook = kwargs.pop('docbook')
+                if not isinstance(docbook, str):
+                    raise MesonException('docbook value must be a string.')
+
+                output = namebase + '-docbook'
+                custom_kwargs = {'input': xml_file,
+                                 'output': output,
+                                 'command': cmd + ['--generate-docbook', docbook, '@INPUT@'],
+                                 'build_by_default': build_by_default
+                                 }
+                targets.append(build.CustomTarget(output, state.subdir, state.subproject, custom_kwargs))
```

Comment:
I'm confused, the `else:` condition here (for older gdbus-codegen) is *also* returning a list of targets? Shouldn't this use the old mechanism of having a single custom target that generates the C source, header, and docbook?

Or did I misunderstand what you decided to do? Because right now the code contradicts the documentation.

Reply:
This is due to a conversation at the IRC, where it was concluded that a version-dependent return was not a good design.

This, along with the fact that for previous versions it can be transparent, since the set of targets can be seen as the opaque target that was returned before, is the reason why I changed the implementation.

Reply:
So maybe we should have one custom target and return it thrice?

Reply:
Because by building things (potentially) thrice, there's a greater chance of things breaking. Returning the same target thrice will satisfy the API requirements and also greatly reduce the chances of regressions for existing projects with older gdbus-codegen.

Reply:
> So maybe we should have one custom target and return it thrice?

The generation of a single target would not fit the goal of these changes, because separate targets are needed to handle dependencies on each target properly.

> Because by building things (potentially) thrice, there's a greater chance of things breaking. Returning the same target thrice will satisfy the API requirements and also greatly reduce the chances of regressions for existing projects with older gdbus-codegen.

Or do you mean splitting this new implementation into three different new functions for each target?